mitxgraders/stringgrader.py

```diff
@@ -29,48 +75,123 @@ def schema_config(self):
+        """
+        cleaned = str(input)
+
+        # Convert \t and newline characters (\r and \n) to spaces
+        # Note: there is no option for this conversion
+        cleaned = cleaned.replace('\t', ' ')
+        cleaned = cleaned.replace('\r\n', ' ')
+        cleaned = cleaned.replace('\n\r', ' ')
+        cleaned = cleaned.replace('\r', ' ')
+        cleaned = cleaned.replace('\n', ' ')
+
+        # Apply case sensitivity
+        if not self.config['case_sensitive']:
+            cleaned = cleaned.lower()
+
+        # Apply strip, strip_all and clean_spaces
+        if self.config['strip']:
+            cleaned = cleaned.strip()
+        if self.config['strip_all']:
+            cleaned = cleaned.replace(' ', '')
+        if self.config['clean_spaces']:
+            cleaned = re.sub(r' +', ' ', cleaned)
+
+        return cleaned
 
     def check_response(self, answer, student_input, **kwargs):
         """
         Grades a student response against a given answer
 
         Arguments:
-            answer (str): The answer to compare to
+            answer (dict): Dictionary describing the expected answer,
+                           its point value, and any associated message
             student_input (str): The student's input passed by edX
         """
-        expect = answer['expect']
-        student = student_input
-
-        # What are we checking?
-        if self.config['accept_any']:
-            correct = True
-        elif self.config['accept_nonempty']:
-            correct = student.strip() != ""
+        expect = self.clean_input(answer['expect'])
+        student = self.clean_input(student_input)
+        incorrect_response = {'ok': False, 'grade_decimal': 0, 'msg': ''}
+
+        # Figure out if we are accepting any input
+        accept_any = self.config['accept_any'] or self.config['accept_nonempty']
+        min_length = self.config['min_length']
+        if self.config['accept_nonempty'] and min_length == 0:
+            min_length = 1
+
+        # Apply the validation pattern
+        pattern = self.config['validation_pattern']
+        if pattern is not None:
+            if not accept_any:
+                # Make sure that expect matches the pattern
+                # If it doesn't, a student can never get this right
+                if re.match(pattern, expect) is None:
+                    msg = "The provided answer '{}' does not match the validation pattern '{}'"
+                    raise ConfigError(msg.format(answer['expect'], pattern))
+
+            # Check to see if the student input matches the validation pattern
+            if re.match(pattern, student) is None:
+                # Either raise an error or grade as incorrect
+                if self.config['validation_err']:
+                    raise InvalidInput(self.config['invalid_msg'])
+                else:
+                    if self.config['debug']:
+                        incorrect_response['msg'] = 'Validation failed'
```

Comment:
(Should have asked this on first review)

Why only add `config[‘invalid_msg’` when raising an error? Why not also show it when returning an incorrect response?

Reply:
You might want to use a regex expression to say "this is the form that the right answer takes" instead of using an expect string. However, you raise a good point - I'll implement the same thing as for minimums.

Reply:
And updated. You'll be really happy to see that I wrote a function!